### PR TITLE
KAFKA-5815: add Printed class and KStream#print(printed)

### PR DESCRIPTION
**Pull request by dguy**
Part of KIP-182
- Add `Printed` class and `KStream#print(Printed)`
- deprecate all other `print` and `writeAsText` methods

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang, @mjsax, @bbejeck - note this is not the complete KAFKA-5652. I'm trying to break it down into multiple smaller PRs so it is manageable and so that i can make progress

**Comment by mjsax** [comment]
Thanks for breaking it down. We should create sub-tasks on the JIRA for each PR thought to make tracking simpler. Thx.

**Comment by dguy** [comment]
Can't make it a sub-task of a sub-task, so i just made it a sub-task of https://issues.apache.org/jira/browse/KAFKA-5651

**Comment by mjsax** [comment]
I see. Too bad.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 1]
```diff
@@ -422,10 +436,18 @@ void print(final Serde<K> keySerde,
...
      * @param valSerde a {@link Serde<V>} used to deserialize value if type is {@code byte[]}.
      * @param label The given name which labels output will be printed.
+     * @deprecated use {@code print(Printed)}
      */
+    @Deprecated
     void print(final KeyValueMapper<? super K, ? super V, String> mapper, final Serde<K> keySerde, final Serde<V> valSerde, final String label);
 
     /**
+     * Print the records of this KStream using the options provided by {@link Printed}
+     * @param printed   options for printing
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: nit: extra space between `printed` and `options`

File: streams/src/test/java/org/apache/kafka/streams/kstream/PrintedTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    public void shouldPrintWithKeyValueMapper() throws UnsupportedEncodingException {
+        final Processor<String, Integer> processor = sysOutPrinter.withKeyValueMapper(new KeyValueMapper<String, Integer, String>() {
+            @Override
+            public String apply(final String key, final Integer value) {
+                return String.format("%s -> %d", key, value);
+            }
+        }).build("processor")
+                .get();
+        processor.process("hello", 1);
+        assertThat(new String(sysOut.toByteArray(), StandardCharsets.UTF_8.name()), equalTo("[processor]: hello -> 1\n"));
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/PrintedTest.java]: nit: replace the duplicate `new String(..)` calls with a private method `getString(byte[])`

**Comment by bbejeck** [comment]
@dguy thanks for PR just a couple minor comments, otherwise LGTM.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+
+import java.io.FileNotFoundException;
+import java.io.PrintWriter;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+
+/**
+ * An object to define the options used when printing a {@link KStream}
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: Nit: `.` missing at the end

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+ * @param <K> key type
+ * @param <V> value type
+ * @see KStream#print(Printed)
+ */
+public class Printed<K, V> {
+    private final PrintWriter printWriter;
+    private String label;
+    private KeyValueMapper<? super K, ? super V, String> mapper = new KeyValueMapper<K, V, String>() {
+        @Override
+        public String apply(K key, V value) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: nit: add `final` ?

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            return String.format("%s, %s", key, value);
+        }
+    };
+
+    private Printed(final PrintWriter printWriter) {
+        this.printWriter = printWriter;
+    }
+
+    /**
+     * Builds the {@link ProcessorSupplier} that will be used to print the records flowing through a {@link KStream}
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: Nit: `.` at the end

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+     *
+     * @return the ProcessorSupplier to be used for printing
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: nit: `{@code ProcessorSupplier}`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 7]
```diff
@@ -233,8 +228,13 @@ public void print(final KeyValueMapper<? super K, ? super V, String> mapper,
                       final String label) {
         Objects.requireNonNull(mapper, "mapper can't be null");
         Objects.requireNonNull(label, "label can't be null");
-        String name = builder.newName(PRINTING_NAME);
-        builder.internalTopologyBuilder.addProcessor(name, new KStreamPrint<>(new PrintForeachAction<>(null, mapper, label), keySerde, valSerde), this.name);
+        print(Printed.<K, V>toSysOut().withLabel(label).withKeyValueMapper(mapper));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:234]: This does not seem to be backward compatible as we ignore `keySerde` and `valSerde` now...

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 8]
```diff
@@ -292,16 +292,7 @@ public void writeAsText(final String filePath,
...
-            throw new TopologyException("filePath can't be an empty string");
-        }
-        final String name = builder.newName(PRINTING_NAME);
-        try {
-            PrintWriter printWriter = new PrintWriter(filePath, StandardCharsets.UTF_8.name());
-            builder.internalTopologyBuilder.addProcessor(name, new KStreamPrint<>(new PrintForeachAction<>(printWriter, mapper, label), keySerde, valSerde), this.name);
-        } catch (FileNotFoundException | UnsupportedEncodingException e) {
-            throw new TopologyException("Unable to write stream to file at [" + filePath + "] " + e.getMessage());
-        }
+        print(Printed.<K, V>toFile(filePath).withKeyValueMapper(mapper).withLabel(label));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:300]: as above. is this backwards compatible?

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+     * @return a new Printed instance
+     */
+    public static <K, V> Printed<K, V> toFile(final String filePath) {
+        Objects.requireNonNull(filePath, "filePath can't be null");
+        if (filePath.trim().isEmpty()) {
+            throw new TopologyException("filePath can't be an empty string");
+        }
+        try {
+            return new Printed<>(new PrintWriter(filePath, StandardCharsets.UTF_8.name()));
+        } catch (FileNotFoundException | UnsupportedEncodingException e) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            throw new TopologyException("Unable to write stream to file at [" + filePath + "] " + e.getMessage());
+        }
+    }
+
+    /**
+     * Print the records of a {@link KStream} to system out
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: Nit: `.`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+     * @return this
+     */
+    public Printed<K, V> withLabel(final String label) {
+        Objects.requireNonNull(label, "label can't be null");
+        this.label = label;
+        return this;
+    }
+
+    /**
+     * Print the records of a {@link KStream} with the provided {@link KeyValueMapper}
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java:102]: Nit: `.`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+     * The example below shows how to customize output data.
+     * <pre>{@code
+     * final KeyValueMapper<Integer, String, String> mapper = new KeyValueMapper<Integer, String, String>() {
+     *     @Override
+     *     public String apply(Integer key, String value) {
+     *         return String.format("(%d, %s)", key, value);
+     *     }
+     * };
+     * }</pre>
+     * <p>
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: nit: not need for `<p>` here as after the `<pre></per>` block there will be space inserted already

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrint.java [hunk 13]
```diff
@@ -20,64 +20,28 @@
...
-import org.apache.kafka.common.serialization.Serde;
 
 public class KStreamPrint<K, V> implements ProcessorSupplier<K, V> {
 
-    private final Serde<?> keySerde;
-    private final Serde<?> valueSerde;
     private final ForeachAction<K, V> action;
     
-    public KStreamPrint(final ForeachAction<K, V> action, final Serde<?> keySerde, final Serde<?> valueSerde) {
+    public KStreamPrint(final ForeachAction<K, V> action) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrint.java:28]: Guess, we need to keep the serdes for now to be backward compatible until we really remove all deprecated methods...

**Review by mjsax — CHANGES_REQUESTED** [review]
Thanks for the PR!

**Comment by mjsax** [comment]
Do we need to update some docs for this? If yes, it should be part of the PR. Same for all other PR for KIP-182

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 7 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:234]: Yep, there was some discussion about it here: https://github.com/apache/kafka/pull/3636#discussion_r131981250 and also offline with @guozhangwang

Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:234]: The issue is that the `key` and `value` are already deserialized into the expected type, i.e., if the stream is `<Integer, String>` then you already have  `Integer key, String value` so the serdes are pointless. You can't pass in another type of `Serde` as they have to be `Serde<K>` and `Serde<V>`

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:234]: Ack. Thanks for clarification.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    private Printed(final PrintWriter printWriter) {
+        this.printWriter = printWriter;
+    }
+
+    /**
+     * Builds the {@link ProcessorSupplier} that will be used to print the records flowing through a {@link KStream}.
+     *
+     * @return the {@code ProcessorSupplier} to be used for printing
+     */
+    public ProcessorSupplier<K, V> build(final String processorName) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: Could we not expose this public API?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 15]
```diff
@@ -29,7 +28,7 @@
 import org.apache.kafka.streams.kstream.KTable;
 import org.apache.kafka.streams.kstream.KeyValueMapper;
 import org.apache.kafka.streams.kstream.Predicate;
-import org.apache.kafka.streams.kstream.PrintForeachAction;
+import org.apache.kafka.streams.kstream.Printed;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:32]: This is not part of the PR but: does `PrintForeachAction` need to be in public package instead of `internals`?

File: docs/streams/developer-guide.html [hunk 16]
```diff
@@ -943,10 +943,14 @@
                 <pre class="brush: java;">
                        KStream&lt;byte[], String&gt; stream = ...;
                        stream.print();
-
-                       // Several variants of `print` exist to e.g. override the default serdes for record keys
-                       // and record values, set a prefix label for the output string, etc
-                       stream.print(Serdes.ByteArray(), Serdes.String());
+
```
Comment (guozhangwang) [inline docs/streams/developer-guide.html:1019]: For doc changes: I think we also need to add a paragraph in upgrade section talking about the removal of `serdes` in print, and that users should not try to use `serdes` anymore but to override the mapper function if their key / values default `toString` function are not good enough (think about `Bytes`).

**Review by guozhangwang — COMMENTED** [review]
Minor comments only.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java (hunk 14 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: Yes, but then we'd need to expose all the other fields. Unless we can think of a way to work around this.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: Could we expose the getters in the sub class, or the build function in the sub class as you mentioned in the other PR?

Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Printed.java]: yeah we can.

**Comment by guozhangwang** [comment]
LGTM!
